extensions/amp-a4a/0.1/amp-ad-render.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {A4AVariableSource} from '../../amp-a4a/0.1/a4a-variable-source';
+import {Renderer} from './amp-ad-type-defs';
+import {createElementWithAttributes} from '../../../src/dom';
+import {dev} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getDefaultBootstrapBaseUrl} from '../../../src/3p-frame';
+import {
+  installFriendlyIframeEmbed,
+  setFriendlyIframeEmbedVisible,
+} from '../../../src/friendly-iframe-embed';
+import {installUrlReplacementsForEmbed} from '../../../src/service/url-replacements-impl';
+import {setStyle} from '../../../src/style';
+
+/**
+ * Render a validated AMP creative directly in the parent page.
+ */
+export class FriendlyFrameRenderer extends Renderer {
+
+  constructor() {
+    super();
+    /** @type {?Element} */
+    this.iframe = null;
+  }
+
+  /** @override */
+  render(context, element, creativeData) {
```

Comment:
I have one question here: from what you implement the context is the AMP.BaseElement instance. However the with the `AmpAdNetworkBase` class, `this.context_` can be any object. What the story here?

Reply:
It wasn't my intention to make `context` an instance of AMP.BaseElement, but the renderer needs to call certain methods like `applyFillContent` and `isInViewport`. My thinking was that those functions would be added to the context object using `context.applyFillContent = this.applyFillContent.bind(this)`, where `this` is an instance of AmpAdNetworkBase. Alternatively, I can pass`this` itself into the context object and call the methods on that.

Reply:
I see. What other baseInstance method are we expecting to be used other than `applyFillContent` and `isInViewport`? does that mean context will always be an `AmpAdNetworkBase` instance for all renderer, if so shall we change the typedef for context?

Reply:
I think those are the only two methods I'll need. @lannka was pretty firm on not passing the `AmpAdNetworkBase` instance to either the validator or the renderer, so I'm _not_ suggesting that `context` be an instance of `AmpAdNetworkBase`, only that we inject these two methods into it and bind them to the `AmpAdNetworkBase` instance, or that we pass the `AmpAdNetworkBase` as a field inside `context`, but that seems no better than just passing the instance.

Reply:
In that case, let try not to pass `this` as the context. 
I suggest we create a API interface that has `applyFillContent` and `isInViewport` if these two functions are always required. Then each context must implement these two functions. What do you think.

Reply:
That sounds good to me. I've added AmpAdContext as an abstract class rather than an interface, because there's some basic functionality that I think should be built in. Take a look and let me know what you think.

Reply:
@glevitzky I am a little bit confused now. 

Question 1:
>  @lannka was pretty firm on not passing the AmpAdNetworkBase instance to either the validator or the renderer, so I'm not suggesting that context be an instance of AmpAdNetworkBase

Do you remember why? 

Question 2 : Should the `context` be overwritten by individual ad integration?

Reply:
1) Not sure exactly what his reasoning was, but here's what he said in an email exchange: "Avoid accessing member fields of the base class in any validator/renderer impl. Instead, a context object will be passed around throughout the ad lifecycle (to validator and renderer etc) to hold states."

2) I don't think `context` would _need_ to be overriden by ad networks in most cases. The main use case of the context object would be to pass arbitrary pieces of data to the renderer like requestUrl, slot size, ampdoc, etc. Initially, we envisioned this as being a simple JSON object. The only complication is these BaseElement instance methods.

Reply:
I realize, however, that the code I have now is basically passing the AmpAdNetworkBase instance anyway. This is why I wanted to keep `context` as a simple object, and let networks that need to use renderers that call `applyFillContent`/`isInViewport` to just do something like `context.applyFillContent = this.applyFillContent.bind(this)`.

Reply:
Right! I agree we shouldn't make `context` extend any class. If it won't be overridden, we don't even need an interface. What you suggest `context.applyFillContent = this.applyFillContent.bind(this)` sounds good. 
Or if we feel these are the only two methods that we need from baseInstance, things can be pretty easy. For example: if you take a look at `applyFillContent` logic, it's simply adding a class to element, we can easily do so w/o using `baseInstance.applyFillContent`.

Reply:
SG. I've reverted the last commit, and will make further updates Monday.

Reply:
Actually, I'm not sure I need to update anything in this PR. We'll make the relevant changes in the AmpAdNetworkBase implementation PRs.